File: Divide_Two_Integers.cpp
/*
Divide two integers without using multiplication, division and mod operator.
If it is overflow, return MAX_INT.

Analysis:
1.当除数divisor为0，x/0是不合法的；
2.INT_MIN=-2147483648,INT_MAX=2147483647，因此-INT_MIN > INT_MAX 不合法；
3.考虑到n<<0=n*2**0,...,n<<1=n*2**1,每做移一位，n会增大一倍，也就是说b<<i=b*2**2，其实就是成倍变化；
3.因此，只需要不断减少divisor的倍数即可，每次都1次开始，直到a<(b<<n)，循环即可。
*/

#include <iostream>
#include <stdlib.h>
#include <limits.h>
using namespace std;


class Solution {
public:
    int divide(int dividend, int divisor)
    {
        int res = 0;

        long a = labs(dividend);
        long b = labs(divisor);

        long sign = (dividend < 0) ^ (divisor < 0) ? -1 : 1;

        if (b == 0 || dividend == INT_MIN && divisor == -1) 
            return INT_MAX;
        else if (b == -1)
            return sign * a;

        int i = 0;
        while (a >= b) {
            i = 0;
            while (a >= b<<i) {
                a -= b<<i;
                res += 1<<i;
                i++;
            }
        }
        
        return res*sign;
    }
};

int main(int argc, char** argv)
{
    Solution a;

    int res = a.divide(5, 4);

    cout<<res<<endl;

    return 1;
}

File: divide-two-integers.py
# -*- coding: utf-8 -*-

"""
# https://leetcode.com/problems/divide-two-integers/

Given two integers dividend and divisor, divide two integers without using multiplication, division and mod operator.

Return the quotient after dividing dividend by divisor.

The integer division should truncate toward zero, which means losing its fractional part.
For example, truncate(8.345) = 8 and truncate(-2.7335) = -2.

Example 1:
Input: dividend = 10, divisor = 3
Output: 3
Explanation: 10/3 = truncate(3.33333..) = 3.

Example 2:
Input: dividend = 7, divisor = -3
Output: -2
Explanation: 7/-3 = truncate(-2.33333..) = -2.

Note:
Both dividend and divisor will be 32-bit signed integers.
The divisor will never be 0.
Assume we are dealing with an environment which could only store integers within the 32-bit signed integer range: [−231,  231 − 1]. For the purpose of this problem, assume that your function returns 231 − 1 when the division result overflows.
"""


class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        x = abs(dividend)
        y = abs(divisor)

        # 32bit int -> [-2147483648, 2147483647]
        if dividend == -2147483648 and divisor == -1:
            return 2147483647
        elif divisor == -1:
            return 0 - dividend
        elif divisor == 1:
            return dividend

        # Method1: Exceed Time
        # i = 0
        # while x >= y:
        #     x -= y
        #     i += 1

        # Method2: Use shift method
        r = 0
        while x >= y:
            i = 0
            while x >= y << i:
                x -= y << i
                r += 1 << i
                i += 1

        if (dividend < 0 < divisor) or (dividend > 0 > divisor):
            return 0 - r

        return r


if __name__ == '__main__':
    # assert Solution().divide(10, 3) == 3
    # assert Solution().divide(7, -3) == -2
    # assert Solution().divide(1, 1) == 1
    assert Solution().divide(-2147483648, 1) == -2147483648
    assert Solution().divide(-2147483648, -1) == 2147483647